File: 백준/Gold/17472. 다리 만들기 2/다리 만들기 2.py
from collections import deque

dx = [-1, 1, 0, 0]
dy = [0, 0, 1, -1]


def bfs(x, y, cnt):
    idx = []
    q = deque()
    q.append((x, y))
    visited[x][y] = cnt
    while q:
        x, y = q.popleft()
        idx.append([x,y])
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if nx >= n or ny >= m or nx < 0 or ny < 0: continue
            if graph[nx][ny] == 1 and not visited[nx][ny]:
                q.append((nx, ny))
                visited[nx][ny] = cnt
    return idx


def findParents(x):
    if x == parents[x]: return x
    parents[x] = findParents(parents[x])
    return parents[x]

def union(aa, bb):
    aRoot = findParents(aa)
    bRoot = findParents(bb)
    if aRoot == bRoot:
        return False
    if aRoot < bRoot:
        parents[bRoot] = aRoot
    else:
        parents[aRoot] = bRoot
    return True

def widthBridgeCheck(i, j):
    cost = 100000
    for l in range(len(island[i])):
        ix = island[i][l][0]
        iy = island[i][l][1]
        for k in range(len(island[j])):
            jx = island[j][k][0]
            jy = island[j][k][1]
            if ix == jx :
                if iy > jy :
                    key = True
                    for r in range(jy+1, iy):
                        if graph[ix][r] == 1:
                            key = False
                            break
                    if key :
                        tmp = abs(jy-iy)-1
                        if tmp < 2: continue
                        cost = min(abs(jy-iy)-1, cost)
                elif iy < jy:
                    key = True
                    for r in range(iy+1, jy):
                        if graph[ix][r] == 1:
                            key = False
                            break
                    if key:
                        tmp = abs(jy - iy) - 1
                        if tmp < 2: continue
                        cost = min(tmp, cost)
    if cost == 100000:
        return
    else:
        edgeList.append([i, j, cost])
        return
def heightBridgeCheck(i, j):
    cost = 100000
    for l in range(len(island[i])):
        ix = island[i][l][0]
        iy = island[i][l][1]
        for k in range(len(island[j])):
            jx = island[j][k][0]
            jy = island[j][k][1]
            if iy == jy:
                if ix > jx:
                    key = True
                    for r in range(jx+1, ix ):
                        if graph[r][iy] == 1:
                            key = False
                            break
                    if key:
                        tmp = abs(jx - ix) - 1
                        if tmp < 2: continue
                        cost = min(tmp, cost)
                elif ix < jx:
                    key = True
                    for r in range(ix+1, jx):
                        if graph[r][iy] == 1:
                            key = False
                            break
                    if key:
                        tmp = abs(jx - ix) - 1
                        if tmp < 2: continue
                        cost = min(tmp, cost)
    if cost == 100000:
        return
    else:
        edgeList.append([i, j, cost])
        return

def make():
    for k in range(len(island) + 1):
        parents[k] = k

n, m = map(int, input().split())
graph = [list(map(int, input().split())) for _ in range(n)]
visited = [[0] * m for _ in range(n)]
island = [[0]]
edgeList = []
cnt = 1
for i in range(n):
    for j in range(m):
        if not visited[i][j] and graph[i][j]:  # bfs를 돌면서 섬들의 시작 인덱스와 끝 인덱스를 기록
            idx = bfs(i, j, cnt)
            island.append(idx)
            cnt += 1

# print(visited)
# print("island : ",island)

for i in range(1, len(island)):
    for j in range(i + 1, len(island)):
        widthBridgeCheck(i, j)  # 가로로 다리 건설이 가능한지 체크. 불가능한 경우 -1 리턴
        heightBridgeCheck(i, j)  # 세로로 다리 건설이 가능한지 체크


parents = [0] * (len(island)+1)
make()
edgeList.sort(key = lambda x: x[2])
cnt = 0
weight = 0
# print("edgelist : ", edgeList)
for edge in edgeList:
    if not union(edge[0], edge[1]):
        continue
    weight += edge[2]
    cnt += 1
    if cnt == len(island)-2:
        break
if cnt != len(island)-2:
    print(-1)
elif weight == 0:
    print(-1)
else:
    print(weight)
# print(weight if weight != 0 else -1)

File: 백준/Gold/17472. 다리 만들기 2/다리 만들기 2.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Deque;

public class Main {

    static int[] dx = {-1, 1, 0, 0};
    static int[] dy = {0, 0, 1, -1};
    static int[][] graph;
    static int[][] visited;
    static ArrayList<ArrayList<int[]>> island;
    static ArrayList<int[]> edgeList;
    static int n, m;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] size = br.readLine().split(" ");
        n = Integer.parseInt(size[0]);
        m = Integer.parseInt(size[1]);

        graph = new int[n][m];
        visited = new int[n][m];
        island = new ArrayList<>();
        island.add(new ArrayList<>());
        edgeList = new ArrayList<>();

        for (int i = 0; i < n; i++) {
            String[] row = br.readLine().split(" ");
            for (int j = 0; j < m; j++) {
                graph[i][j] = Integer.parseInt(row[j]);
            }
        }

        int cnt = 1;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (visited[i][j] == 0 && graph[i][j] == 1) {
                    ArrayList<int[]> idx = bfs(i, j, cnt);
                    island.add(idx);
                    cnt++;
                }
            }
        }

        for (int i = 1; i < island.size(); i++) {
            for (int j = i + 1; j < island.size(); j++) {
                widthBridgeCheck(i, j);
                heightBridgeCheck(i, j);
            }
        }

        int[] parents = new int[island.size() + 1];
        make(parents);
        Collections.sort(edgeList, (a, b) -> a[2] - b[2]);

        int connected = 0;
        int weight = 0;
        for (int[] edge : edgeList) {
            if (!union(parents, edge[0], edge[1])) {
                continue;
            }
            weight += edge[2];
            connected++;
            if (connected == island.size() - 2) {
                break;
            }
        }

        if (connected != island.size() - 2) {
            System.out.println(-1);
        } else if (weight == 0) {
            System.out.println(-1);
        } else {
            System.out.println(weight);
        }
    }

    static ArrayList<int[]> bfs(int x, int y, int cnt) {
        ArrayList<int[]> idx = new ArrayList<>();
        Deque<int[]> q = new ArrayDeque<>();
        q.add(new int[]{x, y});
        visited[x][y] = cnt;

        while (!q.isEmpty()) {
            int[] current = q.poll();
            idx.add(current);
            int cx = current[0];
            int cy = current[1];
            for (int i = 0; i < 4; i++) {
                int nx = cx + dx[i];
                int ny = cy + dy[i];
                if (nx < 0 || ny < 0 || nx >= n || ny >= m) continue;
                if (graph[nx][ny] == 1 && visited[nx][ny] == 0) {
                    q.add(new int[]{nx, ny});
                    visited[nx][ny] = cnt;
                }
            }
        }
        return idx;
    }

    static void widthBridgeCheck(int i, int j) {
        int cost = 100000;
        for (int[] coordI : island.get(i)) {
            int ix = coordI[0];
            int iy = coordI[1];
            for (int[] coordJ : island.get(j)) {
                int jx = coordJ[0];
                int jy = coordJ[1];
                if (ix == jx) {
                    if (iy > jy) {
                        boolean key = true;
                        for (int r = jy + 1; r < iy; r++) {
                            if (graph[ix][r] == 1) {
                                key = false;
                                break;
                            }
                        }
                        if (key) {
                            int tmp = Math.abs(jy - iy) - 1;
                            if (tmp < 2) continue;
                            cost = Math.min(tmp, cost);
                        }
                    } else if (iy < jy) {
                        boolean key = true;
                        for (int r = iy + 1; r < jy; r++) {
                            if (graph[ix][r] == 1) {
                                key = false;
                                break;
                            }
                        }
                        if (key) {
                            int tmp = Math.abs(jy - iy) - 1;
                            if (tmp < 2) continue;
                            cost = Math.min(tmp, cost);
                        }
                    }
                }
            }
        }
        if (cost != 100000) {
            edgeList.add(new int[]{i, j, cost});
        }
    }

    static void heightBridgeCheck(int i, int j) {
        int cost = 100000;
        for (int[] coordI : island.get(i)) {
            int ix = coordI[0];
            int iy = coordI[1];
            for (int[] coordJ : island.get(j)) {
                int jx = coordJ[0];
                int jy = coordJ[1];
                if (iy == jy) {
                    if (ix > jx) {
                        boolean key = true;
                        for (int r = jx + 1; r < ix; r++) {
                            if (graph[r][iy] == 1) {
                                key = false;
                                break;
                            }
                        }
                        if (key) {
                            int tmp = Math.abs(jx - ix) - 1;
                            if (tmp < 2) continue;
                            cost = Math.min(tmp, cost);
                        }
                    } else if (ix < jx) {
                        boolean key = true;
                        for (int r = ix + 1; r < jx; r++) {
                            if (graph[r][iy] == 1) {
                                key = false;
                                break;
                            }
                        }
                        if (key) {
                            int tmp = Math.abs(jx - ix) - 1;
                            if (tmp < 2) continue;
                            cost = Math.min(tmp, cost);
                        }
                    }
                }
            }
        }
        if (cost != 100000) {
            edgeList.add(new int[]{i, j, cost});
        }
    }

    static void make(int[] parents) {
        for (int k = 0; k < island.size() + 1; k++) {
            parents[k] = k;
        }
    }

    static int findParents(int[] parents, int x) {
        if (x == parents[x]) return x;
        parents[x] = findParents(parents, parents[x]);
        return parents[x];
    }

    static boolean union(int[] parents, int aa, int bb) {
        int aRoot = findParents(parents, aa);
        int bRoot = findParents(parents, bb);
        if (aRoot == bRoot) {
            return false;
        }
        if (aRoot < bRoot) {
            parents[bRoot] = aRoot;
        } else {
            parents[aRoot] = bRoot;
        }
        return true;
    }
}